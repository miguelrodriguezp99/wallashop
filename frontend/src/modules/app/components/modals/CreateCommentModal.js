import React, { useState } from "react";

import {
	Button,
	Modal,
	ModalBody,
	ModalContent,
	ModalFooter,
	ModalHeader,
	Textarea,
	useDisclosure,
} from "@nextui-org/react";
import { BiMessageAdd } from "react-icons/bi";
import { toast } from "sonner";
import { createComment, replyComment } from "../../../../backend/postService";

//State puede ser un post o un comentario dependiendo de donde se renderice este componente
export default function CreateCommentModal({ state, reply, setCommentCount }) {
	const { isOpen, onOpen, onOpenChange } = useDisclosure();
	const [text, setText] = useState("");

	const onSuccess = (authenticatedUser) => {
		toast.success("Comment created.");
		setCommentCount((prevCount) => prevCount + 1);
		//navigate(`/wallashop/post/${state.id}`);
		//window.location.reload(true); //RESETEA LA STORE DE ZUSTAND!!!
	};

	const onErrors = (error) => {
		toast.error("Error creating comment.");
	};

	const handleSubmit = async (onClose) => {
		const commentObj = {
			text: text,
		};

		console.log(state);

		reply === false
			? createComment(state.id, commentObj, onSuccess, onErrors)
			: replyComment(state.post.id, commentObj, state.id, onSuccess, onErrors);

		onClose();
	};

	const handleKeyDown = (e, onClose) => {
		if (e.key === "Enter") {
			handleSubmit(onClose);
		}
	}

	return (
		<>
			{reply === false ? (
				<Button
					onPress={() => onOpen()}
					className="border border-gray-600 rounded text-white bg-[#18181B] hover:bg-gray-700 "
				>
					<BiMessageAdd size={25} color="white" />
				</Button>
			) : (
				<Button
					onPress={() => onOpen()}
					className="border border-gray-600 rounded text-white bg-[#18181B] hover:bg-gray-700 "
				>
					Reply
				</Button>
			)}

			<Modal
				isOpen={isOpen}
				size="xl"
				onOpenChange={onOpenChange}
				className="w-full"
			>
				<ModalContent className="bg-[#18181B] w-full">
					{(onClose) => (
						<>
							<ModalHeader className="flex flex-col gap-1 text-white">
								Add Comment
							</ModalHeader>
							<ModalBody className="w-full">
								<Textarea
									labelPlacement="outside"
									placeholder="Enter your description"
									className="w-full"
									value={text}
									onKeyDown={(e) => handleKeyDown(e, onClose)}
									onChange={(e) => setText(e.target.value)}
								/>
							</ModalBody>
							<ModalFooter>
								<Button color="danger" variant="light" onPress={onClose}>
									Close
								</Button>
								<Button color="primary" onPress={() => handleSubmit(onClose())}>
									Submit
								</Button>
							</ModalFooter>
						</>
					)}
				</ModalContent>
			</Modal>
		</>
	);
}
