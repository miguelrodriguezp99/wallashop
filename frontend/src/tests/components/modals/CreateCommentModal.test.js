import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import renderer from 'react-test-renderer';
import CreateCommentModal from '../../../modules/app/components/modals/CreateCommentModal';

describe('CreateCommentModal component', () => {
  it('renders without errors, shows the Submit button, and matches snapshot', () => {
    const state = {
      id: 'postId',
      post: {
        id: 'parentCommentId',
      },
    };
    const setCommentCount = jest.fn();

    const { getByRole } = render(
      <CreateCommentModal state={state} reply={false} setCommentCount={setCommentCount} />
    );

    const submitButton = getByRole('button');
    expect(submitButton).toBeInTheDocument();
    fireEvent.click(submitButton);

    const tree = renderer.create(
      <CreateCommentModal state={state} reply={false} setCommentCount={setCommentCount} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});