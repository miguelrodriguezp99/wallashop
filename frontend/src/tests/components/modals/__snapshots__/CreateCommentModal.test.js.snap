// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateCommentModal component renders without errors, shows the Submit button, and matches snapshot 1`] = `
<button
  className="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent outline-none data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 px-unit-4 min-w-unit-20 h-unit-10 text-small gap-unit-2 [&>svg]:max-w-[theme(spacing.unit-8)] data-[pressed=true]:scale-[0.97] transition-transform-colors-opacity motion-reduce:transition-none data-[hover=true]:opacity-hover border border-gray-600 rounded text-white bg-[#18181B] hover:bg-gray-700"
  disabled={false}
  onBlur={[Function]}
  onClick={[Function]}
  onDragStart={[Function]}
  onFocus={[Function]}
  onKeyDown={[Function]}
  onKeyUp={[Function]}
  onMouseDown={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
  onMouseUp={[Function]}
  onTouchCancel={[Function]}
  onTouchEnd={[Function]}
  onTouchMove={[Function]}
  onTouchStart={[Function]}
  type="button"
>
  <svg
    color="white"
    fill="currentColor"
    height={25}
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": "white",
      }
    }
    viewBox="0 0 24 24"
    width={25}
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M20 2H4c-1.103 0-2 .897-2 2v12c0 1.103.897 2 2 2h3v3.767L13.277 18H20c1.103 0 2-.897 2-2V4c0-1.103-.897-2-2-2zm0 14h-7.277L9 18.233V16H4V4h16v12z"
    />
    <path
      d="M11 14h2v-3h3V9h-3V6h-2v3H8v2h3z"
    />
  </svg>
</button>
`;
